File: Programmers/유석환/Q67256.swift
import Foundation

func solution(_ numbers:[Int], _ hand:String) -> String {
    let location = [(3, 1), (0, 0), (0, 1), (0, 2), (1, 0),
                    (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]
    var left = (3, 0)
    var right = (3, 2)
    var result = ""
    
    for number in numbers {
        switch number {
        case 1, 4, 7:
            left = location[number]
            result += "L"
        case 3, 6, 9:
            right = location[number]
            result += "R"
        default:
            let l = abs(left.0 - location[number].0) + abs(left.1 - location[number].1)
            let r = abs(right.0 - location[number].0) + abs(right.1 - location[number].1)
            
            if l == r {
                if hand == "left" {
                    left = location[number]
                    result += "L"
                } else {
                    right = location[number]
                    result += "R"
                }
            } else if l < r {
                left = location[number]
                result += "L"
            } else {
                right = location[number]
                result += "R"
            }
        }
    }
    
    return result
}

File: Programmers/노수일/Q67256.py
def solution(numbers, hand):
    def distance(finger, num):
        key_pad = [[1, 2, 3], [4, 5, 6], [7, 8, 9], ['*', 0, '#']]
        finger_index = [(i, j) for i in range(4) for j in range(3) if key_pad[i][j] == finger][0]
        num_index = [(i, j) for i in range(4) for j in range(3) if key_pad[i][j] == num][0]

        y_dist = finger_index[0] - num_index[0]
        x_dist = finger_index[1] - num_index[1]
        if x_dist < 0:
            x_dist = x_dist * -1
        if y_dist < 0:
            y_dist = y_dist * -1

        return x_dist + y_dist

    left, right = '*', '#'
    answer = ''

    for number in numbers:
        if number == 0 or number % 3 == 2:
            if distance(left, number) == distance(right, number):
                if hand == 'right':
                    answer += 'R'
                    right = number
                else:
                    answer += 'L'
                    left = number
            else:
                if distance(left, number) > distance(right, number):
                    answer += 'R'
                    right = number
                else:
                    answer += 'L'
                    left = number

        elif number % 3 == 1:
            answer += 'L'
            left = number
        else:
            answer += 'R'
            right = number

    return answer


print(solution([1, 3, 4, 5, 8, 2, 1, 4, 5, 9, 5], 'right'))
print(solution([7, 0, 8, 2, 8, 3, 1, 5, 7, 6, 2], 'left'))
print(solution([1, 2, 3, 4, 5, 6, 7, 8, 9, 0], 'right'))
